### PR TITLE
Add code generation for all wire types for stream encoding

### PR DESCRIPTION
**Pull request by dianale31**
Creates the code generation for all the primitive types, enums, typedefs, structs, and container types. We use the same testing utils as the reader implementation in #496 (hopefully that will reduce rebase conflicts).

### CODE REVIEW

File: gen/benchmark_test.go [hunk 1]
```diff
@@ -9,13 +9,16 @@ import (
...
+	"go.uber.org/thriftrw/protocol/binary"
+	"go.uber.org/thriftrw/protocol/stream"
 	"go.uber.org/thriftrw/ptr"
 	"go.uber.org/thriftrw/wire"
 )
 
 type thriftType interface {
 	ToWire() (wire.Value, error)
 	FromWire(wire.Value) error
+	Encode(stream.Writer) error
```
Comment (witriew) [inline gen/benchmark_test.go:21]: Should be in a separate interface declaration - like you currently have in `gen/utils_for_test.go`

File: gen/field.go [hunk 2]
```diff
@@ -467,6 +471,108 @@ func (f fieldGroupGenerator) FromWire(g Generator) error {
...
+	return g.DeclareFromTemplate(
+		`
+		<$stream := import "go.uber.org/thriftrw/protocol/stream">
+		<$v := newVar "v">
+		<$sw := newVar "sw">
+		// Decode deserializes a <.Name> struct directly from its Thrift-level
+		// representation, without going through an intemediary type.
+		//
+		// An error is returned if a <.Name> struct could not be generated from the wire
+		// representation.
```
Comment (witriew) [inline gen/field.go:484]: Comment :D

File: gen/struct.go [hunk 3]
```diff
@@ -51,6 +51,27 @@ func (s *structGenerator) Reader(g Generator, spec *compile.StructSpec) (string,
...
 
+func (s *structGenerator) Encoder(g Generator, spec *compile.StructSpec) (string, error) {
+	name := encoderFuncName(g, spec)
+	err := g.EnsureDeclared(
+		`
+		<$stream := import "go.uber.org/thriftrw/protocol/stream">
+		<$val := newVar "val">
+		<$sw := newVar "sw">
+		<$t := typeReference .Spec>
+		func <.Name>(<$val> <$t>, <$sw> <$stream>.Writer) error {
```
Comment (witriew) [inline gen/struct.go:62]: I think you should be able to inline the `typeReference .Spec`
```suggestion
		func <.Name>(<$val> <typeReference .Spec>, <$sw> <$stream>.Writer) error {
```

File: gen/field.go [hunk 4]
```diff
@@ -95,6 +95,10 @@ func (f fieldGroupGenerator) Generate(g Generator) error {
 		return err
 	}
 
+	if err := f.Encode(g); err != nil {
+		return err
+	}
```
Comment (witriew) [inline gen/field.go:100]: Don't forget to add "Encode" to the list of `reservedIdentifiers`.

File: protocol/binary/stream_writer.go [hunk 5]
```diff
@@ -65,6 +65,14 @@ func ReturnStreamWriter(sw *StreamWriter) {
 	streamWriterPool.Put(sw)
 }
 
+func NewStreamWriter(w io.Writer) *StreamWriter {
+	return &StreamWriter{
+		Impl:   iface.Impl{},
+		writer: w,
+		buffer: [8]byte{},
+	}
+}
```
[the remark this answers is not in the document]
Comment (dianale31) [reply protocol/binary/stream_writer.go]: Decided to go with the BorrowStreamWriter and ReturnStreamWriter approach which was added in #490

**Pull request closed without merging**